shared-types.d.ts

```diff
@@ -0,0 +1,220 @@
+   *
+   * could be (enter emphasis, enter emphasisMarker, exit emphasisMarker, enter strong, enter strongMarker, exit strongMarker, enter data, exit data)
+   */
+  _events?: Event[]
+
+  /**
+   * This is used for tokens that are already “subtokenized”.
+   *
+   * E.g., when parsing flow, there are content tokens, but those are directly tokenized into definitions/setext/paragraphs
+   */
+  _subevents?: Event[]
+
+  /**
+   * Set to true to mark that a token (e.g., with subevents) is already handled
+   */
+  _contentTokenized?: boolean
+
+  /**
+   * close and open are also used in attention:
+   * depending on the characters before and after sequences (**),
+   * the sequence can open, close, both, or none
+   */
+  _open?: boolean
+
+  /**
+   * close and open are also used in attention:
+   * depending on the characters before and after sequences (**),
+   * the sequence can open, close, both, or none
+   */
+  _close?: boolean
+  _marker?: number
+  _side?: number
+
+  /**
+   * Generally, tabs and spaces behave the same, but in the case of a hard break through trailing spaces ( \n), tabs
+   */
+  _tabs?: boolean
+}
+
+/**
+ *
+ */
+export type Event = [string, Token, unknown]
+
+/**
+ * These these are transitions to update the CommonMark State Machine (CSMS)
+ */
+export interface Effects {
+  /**
+   * Enter and exit define where tokens start and end
+   */
+  enter: (type: Type) => void
+
+  /**
+   * Enter and exit define where tokens start and end
+   */
+  exit: (type: Type) => void
+
+  /**
+   * Consume deals with a character, and moves to the next
+   */
+  consume: (code: number) => void
+
+  /**
+   * Attempt deals with several values, and tries to parse according to those values.
+   * If a value resulted in `ok`, it worked, the tokens that were made are used,
+   * and `returnState` is switched to.
+   * If the result is `nok`, the attempt failed,
+   * so we revert to the original state, and `bogusState` is used.
+   */
+  attempt: (
+    notSureWhatThisIs:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    returnState: unknown,
+    bogusState?: unknown
+  ) => (code: number) => void
+
+  /**
+   * interrupt is used for stuff right after a line of content.
+   */
+  interrupt: (
+    notSureWhatThisIs:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: number) => void
+
+  check: (
+    notSureWhatThisIs:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: number) => void
+
+  /**
+   * lazy is used for lines that were not properly preceded by the container.
+   */
+  lazy: (notSureWhatThisIs:
+    | Construct
+    | Construct[]
+    | {[code: number]: Construct | Construct[]},
+  ok: Okay,
+  nok?: NotOkay) => void
+}
+
+/**
+ *
+ */
+export type Okay = (code: number) => () => void
```

Comment:
Does the inner function have any parameters?

Reply:
Should the params in check/lazy/interrupt also be called “returnState” and “bogusState”.

The tokenizer function gets `effects, ok, nok`, which is kind of like the executor function given to Promise: `resolve, reject`.
On the outside, you can pass `returnState, bogusState`, which is kind of like the functions given to `then`: `onfulfill, onreject`

They all get character codes

Reply:
All these functions also should return another function: the next state-as-a-function to go to.
But there is one case where they return void: for the eof character code (at the end of a value)
The reason being: well, there isn’t any state that makes sense, so void works well. Practically that has also helped: if for some reason it was a mistake, then an exception is throw because there is no next function, meaning it surfaces early.